### PR TITLE
GCW-2193 Order priority rules

### PR DESCRIPTION
**Pull request by patrixr**
The rules we're trying to match are described here : https://jira.crossroads.org.hk/browse/GCW-2193

### CODE REVIEW

File: app/models/order.rb [hunk 1]
```diff
@@ -89,6 +89,25 @@ def set_initial_state
     self.state ||= :draft
   end
 
+  def is_priority?
+    now = Time.now.in_time_zone
+    last_6pm = now.change(hour: 18, min: 0, sec: 0)
+    last_6pm -= 24.hours if now < last_6pm
+
+    case state
+    when "submitted" then
```
Comment (namrataukirde) [inline app/models/order.rb]: then is not needed here. we can remove all `then`.

File: app/models/order.rb [hunk 2]
```diff
@@ -89,6 +89,25 @@ def set_initial_state
     self.state ||= :draft
   end
 
+  def is_priority?
+    now = Time.now.in_time_zone
+    last_6pm = now.change(hour: 18, min: 0, sec: 0)
+    last_6pm -= 24.hours if now < last_6pm
```
Comment (namrataukirde) [inline app/models/order.rb]: Another way for this.

`last_pm = last_6pm.advance(days: -1) if now < last_6pm`

not necessary to implement.

File: app/models/order.rb [hunk 3]
```diff
@@ -89,6 +89,25 @@ def set_initial_state
...
+
+    case state
+    when "submitted" then
+      submitted_at && submitted_at <= now - 24.hours
```
Comment (namrataukirde) [inline app/models/order.rb]: Another way : 

`submitted_at <= now.advance(days: -1)`

File: spec/models/order_spec.rb [hunk 4]
```diff
@@ -63,6 +63,96 @@
...
+        old_order = create :order, state: "submitted", submitted_at: Time.now - 25.hours
+        order = create :order, state: "submitted", submitted_at: Time.now - 23.hours
+        expect(old_order.is_priority?).to eq(true)
+        expect(order.is_priority?).to eq(false)
+      end
+    end
+
+    context 'An order under review (aka processing)' do
+      
+      after { allow(Time).to receive(:now).and_call_original } # un-stub
```
Comment (namrataukirde) [inline spec/models/order_spec.rb]: We have `timecop` gem added. That can be used here.

https://github.com/crossroads/api.goodcity/blob/master/Gemfile#L91

Documentation: https://github.com/travisjeffery/timecop

File: app/models/order.rb [hunk 5]
```diff
@@ -89,6 +89,25 @@ def set_initial_state
     self.state ||= :draft
   end
 
+  def is_priority?
```
Comment (namrataukirde) [inline app/models/order.rb]: Just suggestion: This method name can be changed to `priority?` instead of `is_priority?`

Comment (namrataukirde) [inline app/models/order.rb]: Cyclomatic compexity of this method is high ? Can we do something ?

File: spec/models/order_spec.rb (hunk 4 above)
Comment (patrixr) [reply spec/models/order_spec.rb]: that's pretty cool

File: spec/models/order_spec.rb [hunk 6]
```diff
@@ -63,6 +63,96 @@
...
+
+      context 'If we\'re past 6pm' do
+        before { allow(Time).to receive(:now).and_return(after_6pm_today) }
+
+        it 'should be prioritised if process was started before 6pm and hasn\'t finished' do
+          order_started_before_6 = create :order, state: "processing", processed_at: before_6pm_today
```
Comment (namrataukirde) [inline spec/models/order_spec.rb]: Again just a suggestion you can add trait here in order factory.

create :order, state: "processing", processed_at: before_6pm_today

As this is repeated at multiple places.

File: app/models/order.rb (hunk 5 above)
Comment (patrixr) [reply app/models/order.rb]: @namrataukirde 
It's high complexity from a code processor point of view, for the human eye it isn't. 
Trying to please code climate would just make it worse for us.
I could make a small rule engine, pseudo-code :
```javascript
const priority_rules = {
'submitted': (order) => { /*write rule */ }
};

// in the order
is_priority() {
  return priority_rules[order.state](order);
}
```
I don't think the switch case is that much of a pain, but I'm okay with changing it

Comment (patrixr) [reply app/models/order.rb]: I kinda like is_priority bettrer 😢

Comment (namrataukirde) [reply app/models/order.rb]: Yeah, Switch case is good here. Dont want to remove switch case here. 

If we can move last_6pm logic to some method then that can work.

But whatever we have now is more readable. So not in completely favour of moving code somewhere else.

Just wanted to know everyone's thoughts.

Comment (namrataukirde) [reply app/models/order.rb]: ohh Okay, Then keep it. No worries.

File: app/models/order.rb (hunk 1 above)
Comment (patrixr) [reply app/models/order.rb]: Done

File: spec/models/order_spec.rb [hunk 7]
```diff
@@ -63,6 +63,96 @@
...
+    context 'An order awaiting dispatch' do
+      let(:transport_before_6) { create :order_transport, scheduled_at: before_6pm_today, timeslot: "3PM" }
+      let(:transport_after_6) { create :order_transport, scheduled_at: after_6pm_today, timeslot: "19PM" }
+
+      before { Timecop.freeze(at_6pm_today) }
+
+      it 'should be prioritised if we\'re past it\'s planned dispatch schedule' do
+        priority_order = create :order, state: "awaiting_dispatch", order_transport: transport_before_6
+        non_priority_order = create :order, state: "awaiting_dispatch", order_transport: transport_after_6
+        expect(priority_order.is_priority?).to eq(true)
```
Comment (namrataukirde) [inline spec/models/order_spec.rb]: expect(priority_order.is_priority?).to be_truthy  acn also be used.

Comment (patrixr) [reply spec/models/order_spec.rb]: I like it 👍

Comment (patrixr) [reply spec/models/order_spec.rb]: Although on second thought, truthy does not enforce a type. I'd rather make sure a boolean is returned.
@namrataukirde does the '?' prefix of the method enforce the type to boolean ? or is it just a convention ?

File: app/models/order.rb [hunk 8]
```diff
@@ -336,4 +353,11 @@ def order
   def offer
     nil
   end
+
+  def last_end_of_work_day
+    now = Time.now.in_time_zone
+    last_6pm = now.change(hour: 18, min: 0, sec: 0)
```
Comment (steveyken) [inline app/models/order.rb]: I think you can do Time.now.yesterday and use that as a base. But this works too.

File: spec/models/order_spec.rb (hunk 7 above)
Comment (namrataukirde) [reply spec/models/order_spec.rb]: It's just and convention.

**Pull request merged**